ssl/ssl_lib.c

```diff
@@ -4118,6 +4171,43 @@ SSL_CTX *SSL_CTX_new_ex(OSSL_LIB_CTX *libctx, const char *propq,
         goto err;
     }
 
+#ifndef OPENSSL_NO_SSLKEYLOG
+    if (keylogfile != NULL && strlen(keylogfile) != 0) {
+        /* Make sure we have a global lock allocated */
+        if (!RUN_ONCE(&ssl_keylog_once, ssl_keylog_init)) {
+            /* log an error, but don't fail here */
+            ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
+            goto err;
+        }
+
+        /* Grab out global lock */
+        if (!CRYPTO_THREAD_write_lock(keylog_lock)) {
+            ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
+        } else {
+            /*
+             * If the bio for the requested keylog file hasn't been
+             * created yet, go ahead and create it, and set it to append
+             * if its already there.
+             */
+            if (keylog_bio == NULL) {
+                keylog_bio = BIO_new_file(keylogfile, "a");
+                if (keylog_bio == NULL)
+                    ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
+                else
+                    /* up our ref count for the newly created case */
+                    keylog_count++;
+            } else {
+                /* up our refcount for the already-created case */
+                keylog_count++;
+            }
+            /* If we have a bio now, assign the callback handler */
+            if (keylog_bio != NULL)
+                ret->sslkeylog_callback = sslkeylogfile_cb;
```

Comment:
We only ever set this to `sslkeylogfile_cb` or it could be NULL. It seems overkill to store this in the `SSL_CTX`. Why not just decide whether or not to call it based on the presence of the BIO and call it directly if needed?

Reply:
That was done for the case in which an application might do:
```
setenv("SSLKEYLOGFILE", "/path/to/log/file");
SSL_CTX_new()
setenv("SSLKEYLOGFILE", NULL)
SSL_CTX_new()
```
I.e. a case in which a user only wants to record the keylog data for a specific ctx.  Not sure how common that would be, but that was the intention.

Reply:
Not sure why that means you need to store the callback. It's only purpose is a flag to indicate whether to log to a file or not.

Reply:
Also, I'm not really convinced yet that doing this at the SSL_CTX level is the right place. It kind of feels to me like this might be more appropriate at the SSL level. Most frequently I think you're only going to want to do this for a single connection.

Reply:
Well, the intent was to allow some contexts to be unset while others were set.  Yes its effectively a flag, but its the difference between setting a callback, or setting a flag and making a call.  Not sure theres alot of advantage one way or the other there.

As for the use of SSL vs SSL_CTX, I chose the latter because thats where keylogging is already implemented via SSL_CTX_set_keylog_callback, so it seems a natural fit already.

Reply:
> As for the use of SSL vs SSL_CTX, I chose the latter because thats where keylogging is already implemented via SSL_CTX_set_keylog_callback, so it seems a natural fit already.

But the draft is clearly worded in the singular (my emphasis added): "supports the logging information about the secrets used in **a** TLS connection". It doesn't seem like it would be a well-formed file if secrets from multiple different connections were added to it. If we have multiple different SSL connections started in parallel all using the same SSL_CTX then we are going to end up with an unusable mess. Even if we have multiple connections started in series, we are going to end up with multiple sets of secrets in a single file - which also does not seem well formed.

Reply:
I don't see how that would be the case, each entry in the sslkeylog file is indexed by the connections client random value, so independent connections can be interleaved within the same file.

Reply:
> Well, the intent was to allow some contexts to be unset while others were set. Yes its effectively a flag, but its the difference between setting a callback, or setting a flag and making a call. Not sure theres alot of advantage one way or the other there.

The disadvantage is the unnecessary indirection when reading the code. When reading the code where the callback is called, you have to go look elsewhere in the code to figure out what the callback value can be. If it was just called directly this wouldn't be necessary, i.e. so it makes things more complicated than they need to be.

Reply:
Though I understand the desire to make the documentation reflect the potential to record multiple sessions in parallel

Reply:
> Additionally, while the draft indicates singular use, it also explicitly indicates that multiple connections can be logged in parallel:

You're right! I had not spotted that.

Reply:
It's completely normal to have multiple connections in one file. I do this all the time.